apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/logging/core/PatternLogger.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright 2011 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+package org.apache.skywalking.apm.agent.core.logging.core;
+
+import org.apache.skywalking.apm.agent.core.conf.Config;
+import org.apache.skywalking.apm.util.PropertyPlaceholderHelper;
+import org.apache.skywalking.apm.util.StringUtil;
+
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import java.util.Properties;
+
+/**
+ * @author alvin
+ */
+public class PatternLogger extends EasyLogger {
+
+    public static final String DEFAULT_PATTERN = "%{level} %{timestamp} %{thread} %{class} : %{msg} %{throwable}";
+
+    private String pattern;
+
+    public PatternLogger(Class targetClass, String pattern) {
+        this(targetClass.getSimpleName(), pattern);
+    }
+
+    public PatternLogger(String targetClass, String pattern) {
+        super(targetClass);
+        this.setPattern(pattern);
+    }
+
+    public String getPattern() {
+        return pattern;
+    }
+
+    public void setPattern(String pattern) {
+        if (StringUtil.isEmpty(pattern)) {
+            pattern = DEFAULT_PATTERN;
+        }
+        this.pattern = pattern;
+    }
+
+    @Override
+    String format(LogLevel level, String message, Throwable t) {
+        Properties props = buildContext(level, message, t);
+        return PropertyPlaceholderHelper.LOGGER.replacePlaceholders(getPattern(), props);
+    }
+
+    private Properties buildContext(LogLevel level, String message, Throwable t) {
```

Comment:
You should test performance by the benchmark. @kezhenxu94 Do you remember a PR to give an example?

Reply:
Yes, @mm23504570 please refer to https://github.com/apache/skywalking/pull/2766#issuecomment-496240702 for how to write a benchmark show the performance result